Doc(eos_designs): Dual DC example (#2326)
aristanetworks/ansible-avd: ansible_collections/arista/avd/examples/dual-dc-l3ls/README.md

changed: [localhost]

PLAY RECAP
****************************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

After the playbook has run successfully, the directory structure will look as shown below, the contents of which will be covered in later sections:

```shell
ansible-avd-examples/ (or wherever the playbook was run)
  |── dual-dc-l3ls
    ├── ansible.cfg
    ├── documentation
    ├── group_vars
    ├── images
    ├── intended
    ├── inventory.yml
    ├── playbook.yml
    ├── README.md
    └── switch-basic-configurations
```

!!! info
    If the content of any file is ***modified*** and the playbook is rerun, the file ***will not*** be overwritten. However, if any file in the example is ***deleted*** and the playbook is rerun, Ansible will re-create the file.

## Overall design overview

### Physical topology

> The drawing below shows the physical topology used in this example. The interface assignment is identical to the single DC example. The only change introduced in this topology is the DCI links, which are connected to port Ethernet 6 in dc1-leaf2a/b to dc2-leafa/b:

Reviewer: Small typo here at the end of the sentence, dc2-leaf missing numbering.

# AVD example for a dual data center using L3LS

## Introduction

> This example is meant to be used as the logical evolution in AVD to new users. This document follows the single data center structure and asumes the reader was already been able to successfully deploy the scenario described in the section [AVD example for a single data center using L3LS](../single-dc-l3ls/README.md).

Reviewer 2: ```suggestion
This example is meant to be used as the logical evolution in AVD to new users. Therefore, this document follows the single data center structure and assumes the reader could successfully deploy the scenario described in the section [AVD example for a single data center using L3LS](../single-dc-l3ls/README.md).
```

> The main goal of this section is to provide a base scenario to enable the new users to grow in number of DCs (Data Center) and configure the interconnection between different DCs.

Reviewer 2: ```suggestion
The main goal of this example is to provide a base scenario to enable the new users to grow in the number of DCs (Data Centers) and configure the interconnection between different DCs.
```

> The example includes and describes all the AVD files and their content used to build an L3LS EVPN/VXLAN Symmetric IRB network covering two twin DCs. This design will use the EVPN DC GW feature to avoid unnecessary BGP overlay convergence. This feature rewrites EVPN routes on a set of devices (called EVPN Gateways) and shows the Gateways the as next hop for advertisements outside of the EVPN domain using a DCI (DC Interconnect).

Reviewer 2: ```suggestion
The example includes and describes all the AVD files and their content used to build an L3LS EVPN/VXLAN Symmetric IRB network covering two twin DCs. This design will use the EVPN DC GW feature to avoid unnecessary BGP overlay convergence. This feature rewrites EVPN routes on a set of devices (called EVPN Gateways) and shows the Gateways as the next hop for advertisements outside the EVPN domain using a DCI (DC Interconnect).
```

> The example includes and describes all the AVD files and their content used to build an L3LS EVPN/VXLAN Symmetric IRB network covering two twin DCs. This design will use the EVPN DC GW feature to avoid unnecessary BGP overlay convergence. This feature rewrites EVPN routes on a set of devices (called EVPN Gateways) and show the Gateways as next hop for advertisements outside of the EVPN domain using a DCI (DC Interconnect).

Reviewer 2: ```suggestion
The example includes and describes all the AVD files and their content used to build an L3LS EVPN/VXLAN Symmetric IRB network covering two twin DCs. This design will use the EVPN DC GW feature to avoid unnecessary BGP overlay convergence. This feature rewrites EVPN routes on a set of devices (called EVPN Gateways) and shows the Gateways as the next hop for advertisements outside the EVPN domain using a DCI (DC Interconnect).
```

### DC1

- Two (virtual) spine switches.
- Two sets of (virtual) leaf switches, serving endpoints such as servers.
- Two (virtual) layer2-only switches often used for management connectivity to the servers.

### DC2

- Two (virtual) spine switches.
- Two sets of (virtual) leaf switches, serving endpoints such as servers.
- Two (virtual) layer2-only switches often used for management connectivity to the servers.

In order to follow the principle of network design using patterns, both DCs have the same layout to keep a repetitive pattern using the Leaf and Spine architecture.

> There is, however, no hard requirement to have the exact same number of devices. As an example, one DC could use 2 spines while the other could be using 4 spines to comply with redundancy and bandwidth requirements.

Reviewer 2: ```suggestion
There is, however, no hard requirement to have the exact same number of devices. For example, one DC could use 2 spines while the other could use 4 spines to comply with redundancy and bandwidth requirements.
```

changed: [localhost]

PLAY RECAP
****************************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

After the playbook has run successfully, the directory structure will look as shown below, the contents of which will be covered in later sections:

```shell
ansible-avd-examples/ (or wherever the playbook was run)
  |── dual-dc-l3ls
    ├── ansible.cfg
    ├── documentation
    ├── group_vars
    ├── images
    ├── intended
    ├── inventory.yml
    ├── playbook.yml
    ├── README.md
    └── switch-basic-configurations
```

!!! info
    If the content of any file is ***modified*** and the playbook is rerun, the file ***will not*** be overwritten. However, if any file in the example is ***deleted*** and the playbook is rerun, Ansible will re-create the file.

## Overall design overview

### Physical topology

> The drawing below shows the physical topology used in this example. The interface assignment is identical as the single DC example. The only change introduced in this topology is the DCI links, which are connected to port Ethernet 6 in leaves number two in each DC connecting to the twin DC:

Reviewer 2: ```suggestion
The drawing below shows the physical topology used in this example. The interface assignment is identical to the single DC example. The only change introduced in this topology is the DCI links, which are connected to port Ethernet 6 in dc1-leaf2a/b to dc2-leafa/b:
```

> ![Figure: Arista Leaf Spine physical topology](images/avd-dual-dc-l3ls-example.svg)

Reviewer 3: Typo in the device naming for spines in DC2

| DC1                                                 | 10.255.1.0/27               |
| DC2                                                 | 10.255.129.0/27             |
| **VTEP Loopbacks used for diagnostics**             | **(Leaf switches)**         |
| VRF10 - Common for both DCs                         | 10.255.10.0/27              |
| VRF11 - Common for both DCs                         | 10.255.11.0/27              |
| **SVIs (interface vlan...)**                        | **10.10.`<VLAN-ID>`.0/24**  |
| For example `interface VLAN11` has the IP address:  | 10.10.11.1                  |
| **MLAG Peer-link (interface vlan 4094)**            | **(Leaf switches)**         |
| DC1                                                 | 10.255.1.64/27              |
| DC2                                                 | 10.255.129.64/27            |
| **MLAG iBGP Peering (interface vlan 4093)**         | **(Leaf switches)**         |
| DC1                                                 | 10.255.1.96/27              |
| DC2                                                 | 10.255.129.96/27            |

### BGP design

> Note that new BGP connections are established from dc1-leaf2a and dc1-leaf2b to dc2-leaf2a and dc2-leaf2b respectively. All BGP learnt routes are redistributed into the underlay. This allows for an easier troubleshooting, as all router IDs (Loopback0 IP addresses) and VTEPs (Loopback1 IP addresses) are reachable at a fabric level.

Reviewer 2: ```suggestion
New BGP connections are established from dc1-leaf2a and dc1-leaf2b to dc2-leaf2a and dc2-leaf2b, respectively. In addition, all BGP learned routes are redistributed into the underlay. This allows for easier troubleshooting, as all router IDs (Loopback0 IP addresses) and VTEPs (Loopback1 IP addresses) are reachable at a fabric level.
```

> From the overlay perspective, new each leaf sees its peer in the twin DC as a new Route Server and will advertise all EVPN learnt routes as they were directly connected locally, making all changes inside its DC transparent to the twin DC.

Reviewer 2: ```suggestion
From the overlay perspective, each new leaf sees its peer in the twin DC as a new Route Server and will advertise all EVPN learned routes as they were directly connected locally, making all changes inside its DC transparent to the twin DC.
```

=== "Underlay"

>   ![Figure: Arista Underlay BGP Design](images/bgp-underlay.svg)

Reviewer 2: ```suggestion
    ![Figure: Arista Underlay BGP Design](images/bgp-underlay.svg)
```

Reviewer 2: indent more to get underlay and overlay in the same content tab

>     ![Figure: Arista Underlay BGP Design](images/bgp-underlay.svg)

Reviewer 3: Figure seems to be missing?

Author: incorrect tab spacing

=== "Overlay"

>   ![Figure: Arista Overlay BGP Design](images/bgp-overlay.svg)

Reviewer 2: ```suggestion
    ![Figure: Arista Overlay BGP Design](images/bgp-overlay.svg)
```

>     ![Figure: Arista Underlay BGP Design](images/bgp-overlay.svg)

Reviewer 3: Figure seems to be missing?

Author: same as above, incorrect tab

### Basic EOS config

As discussed in the single DC example, basic connectivity between the Ansible host and the switches must be established before Ansible can be used to push configurations. Remember, you must configure the following on all switches:

- A hostname configured purely for ease of understanding.
- An IP enabled interface - in this example the dedicated out-of-band management interface is used.
- A username and password with the proper access privileges.

!!! note
    The folder `dual-dc-l3ls/switch-basic-configurations/` contains a file per device for the initial configurations.

## Ansible inventory, group vars and naming scheme

The following drawing shows a graphic overview of the Ansible inventory, group variables, and naming scheme used in this example:

![Figure: Ansible inventory and vars](images/ansible-groups.svg)

!!! note
    The two servers on DC1 (`dc1-leaf1-server1` and `dc1-leaf2-server1`) and the two servers on DC2 (`dc2-leaf1-server1` and `dc2-leaf2-server1`) at the bottom are **not** configured by AVD, but the switch ports used to connect to the servers are.

> The same pattern used in the single DC example is followed, where

Reviewer 2: ```suggestion
The same pattern used in the single DC example is followed.
```

- Group names use uppercase and underscore
- All hostnames use lowercase and dashes

> The drawing also shows the relationships between groups and their children. Be aware that all declarations on a higher level are inheritet to chidren automatically.

Reviewer 2: ```suggestion
The drawing also shows the relationships between groups and their children. Be aware that all declarations on a higher level are inherited by children automatically.
```

### Content of the inventory.yml file

This section describes the entire `ansible-avd-examples/dual-dc-l3ls/inventory.yml` file used to represent the above topology.

> In this example, we will consider that no DNS is available and define the IPs that the Ansible host has to reach per device.

Reviewer 3: I can see from a virtual example point of view that having all management IPs in a single contiguous subnet is easier, but for a more real-world scenario, wouldn't it make sense to use dedicated / separate subnets?

Author: done

> In this example we will consider that no DNS is avaliable and will define the IPs that the Ansible host hast to reach per device.

Reviewer 2: ```suggestion
In this example, we will consider that no DNS is available and define the IPs the Ansible host has to reach per device.
```

                  ansible_host: 172.16.1.21
                dc2-spine2:
                  ansible_host: 172.16.1.22
            DC2_L3_LEAVES:
              hosts:
                dc2-leaf1a:
                  ansible_host: 172.16.1.201
                dc2-leaf1b:
                  ansible_host: 172.16.1.202
                dc2-leaf2a:
                  ansible_host: 172.16.1.203
                dc2-leaf2b:
                  ansible_host: 172.16.1.204
            DC2_L2_LEAVES:
              hosts:
                dc2-leaf1c:
                  ansible_host: 172.16.1.251
                dc2-leaf2c:
                  ansible_host: 172.16.1.252

        NETWORK_SERVICES:
          children:
            DC1_L3_LEAVES:
            DC1_L2_LEAVES:
            DC2_L3_LEAVES:
            DC2_L2_LEAVES:
        CONNECTED_ENDPOINTS:
          children:
            DC1_L3_LEAVES:
            DC1_L2_LEAVES:
            DC2_L3_LEAVES:
            DC2_L2_LEAVES:
```

## Defining device types

As discussed in the single DC scenario, all device types must be explicitly defined. The device types are specified in the group variable files with the following names and content:

=== "DC1_SPINES.yml"

> 

Reviewer 3: Something weird is going on with the rendering it seems?

    ```yaml
    --8<--
    examples/single-dc-l3ls/group_vars/DC1_SPINES.yml
    --8<--
    ```

=== "DC1_L3_LEAVES.yml"

    ```yaml
    --8<--
    examples/single-dc-l3ls/group_vars/DC1_L3_LEAVES.yml
    --8<--
    ```

=== "DC1_L2_LEAVES.yml"

    ```yaml
    --8<--
    examples/single-dc-l3ls/group_vars/DC1_L2_LEAVES.yml
    --8<--
    ```

=== "DC2_SPINES.yml"

    ```yaml
    --8<--
    examples/dual-dc-l3ls/group_vars/DC2_SPINES.yml
    --8<--
    ```

=== "DC2_L3_LEAVES.yml"

    ```yaml
    --8<--
    examples/dual-dc-l3ls/group_vars/DC2_L3_LEAVES.yml
    --8<--
    ```

=== "DC2_L2_LEAVES.yml"

    ```yaml
    --8<--
    examples/dual-dc-l3ls/group_vars/DC2_L2_LEAVES.yml
    --8<--
    ```

## Setting fabric-wide configuration parameters

The `ansible-avd-examples/dual-dc-l3ls/group_vars/FABRIC.yml` file defines generic settings that apply to all children of the `FABRIC` group as specified in the inventory described earlier.

> In this section, only additions to previous example will be discussed:

Reviewer 3: only additions to *the* previous example

```yaml title="FABRIC.yml"
l3_edge:
  p2p_links_ip_pools:
    DCI_IP_pool: 172.100.100.0/24 # (1)!
  p2p_links_profiles:
    DCI_profile: # (2)!
      ip_pool: DCI_IP_pool
      as: [65102,65202]
      include_in_underlay_protocol: true
  p2p_links:
  - id: 1 # (3)!
    nodes: [dc1-leaf2a,dc2-leaf2a]
    interfaces: [Ethernet6,Ethernet6]
    profile: DCI_profile
  - id: 2
    nodes: [dc1-leaf2b,dc2-leaf2b]
    interfaces: [Ethernet6,Ethernet6]
    profile: DCI_profile
```

1. Define a new IP Pool, with the name "DCI_IP_pool" which will use the prefix "172.100.100.0/24" to assign IP addresses.
2. Define a new link profile, called "DCI_profile" which will use the previously created IP pool, the ASN will be "65102" for left devices and "65202" for right devices. And interfaces will be included in the underlay routing protocol.

> 3. Define a new link, with the left and right node hostname defined in AVD, along with the interface and finally assign a profile, which contains all required information to configure the link.

Reviewer 2: ```suggestion
3. Define a new link, with the left and right node hostname defined in AVD, along with the interface, and finally assign a profile containing all required information to configure the link.
```

## Setting device specific configuration parameters

> Both `ansible-avd-examples/dual-dc-l3ls/group_vars/DC1.yml` and `ansible-avd-examples/dual-dc-l3ls/group_vars/DC2.yml` files defines settings that apply to all children of `DC1` and `DC2` groups as specified in the inventory described earlier.

Reviewer 2: ```suggestion
Both `ansible-avd-examples/dual-dc-l3ls/group_vars/DC1.yml` and `ansible-avd-examples/dual-dc-l3ls/group_vars/DC2.yml` files define settings that apply to all children of `DC1` and `DC2` groups as specified in the inventory described earlier.
```

> In this section we are going to discuss `DC2` as `DC1` is the same as discussed in the previous example. The only change, the one used to include the EVPN DC GW will be discussed later.

Reviewer 2: ```suggestion
In this section, we will discuss `DC2` as `DC1` is the same as shown in the previous example. The only change, the one used to include the EVPN DC GW will be discussed later.
```

However, this time the settings defined are no longer fabric-wide but are limited to DC2. In this example, DC2 follows the same configuration as DC1, but we will be able to modify the behavior of DC2 without impacting DC1.

> ```yaml title="DC2.yml"

Reviewer 3: I think the title/filename should be clearer.

> However, this time the settings defined are no longer fabric-wide but are limited to DC2. In this example, DC2 follows the same configuration as DC1, but we will be able to modify the behaviour of DC2 without impacting DC1.

Reviewer 2: ```suggestion
However, this time the settings defined are no longer fabric-wide but are limited to DC2. In this example, DC2 follows the same configuration as DC1, but we will be able to modify the behavior of DC2 without impacting DC1.
```

      mgmt_ip: 172.16.1.22/24
```

1. The default gateway for the management interface of all devices in DC2 is defined.
2. `platform` references default settings defined in AVD specific to certain switch platforms.
3. `loopback_ipv4_pool` defines the IP scope from which AVD assigns IPv4 addresses for Loopback0.
4. `bgp_as` defines the BGP AS number.
5. `bgp_defaults` defines generic BGP settings.
6. `nodes` defines the actual spine switches, using the hostnames defined in the inventory.
7. `id` is used to calculate the various IP addresses, for example, the IPv4 address for the Loopback0 interface. In this case, dc2-spine1 will get the IPv4 address 10.255.128.11/27 assigned to the Loopback0 interface.
8. `mgmt_ip` defines the IPv4 address of the management interface. As stated earlier, Ansible will perform name lookups using the hostnames specified in the inventory unless using the `ansible_host` option. However, there is no automatic mechanism to grab the result of the name lookup and use that to generate the management interface configuration.

The following section covers the L3 leaf switches. Significantly more settings need to be set compared to the spine switches:

> ```yaml title="DC2.yml"

Reviewer 3: Again, the title/filename is not 100% clear.

  node_groups: # (14)!
    DC2_L3_LEAF1:
      bgp_as: 65201 # (15)!
      nodes:
        dc2-leaf1a:
          id: 11
          mgmt_ip: 172.16.1.201/24
          uplink_switch_interfaces: # (16)!
            - Ethernet1
            - Ethernet1
        dc2-leaf1b:
          id: 12
          mgmt_ip: 172.16.1.202/24
          uplink_switch_interfaces:
            - Ethernet2
            - Ethernet2

    DC2_L3_LEAF2:
      bgp_as: 65202
      evpn_gateway: # (17)!
        evpn_l2:
          enabled: true
        evpn_l3:
          enabled: true
          inter_domain: true
      nodes:
        dc2-leaf2a:
          id: 13
          mgmt_ip: 172.16.1.203/24
          uplink_switch_interfaces:
            - Ethernet3
            - Ethernet3
          evpn_gateway:
            remote_peers: # (18)!
              - hostname: dc1-leaf2a
        dc2-leaf2b:
          id: 14
          mgmt_ip: 172.16.1.204/24
          uplink_switch_interfaces:
            - Ethernet4
            - Ethernet4
          evpn_gateway:
            remote_peers:
              - hostname: dc1-leaf2b
```

1. `platform` references default settings defined in AVD specific to certain switch platforms.

> 2. `loopback_ipv4_pool` defines the IP scope from which AVD assigns IPv4 addresses for Loopback0. Please note that this IP pool is identical to the one used for the spine switches in this example. To avoid setting the same IP addresses for several devices, we define the option `loopback_ipv4_offset`.

Reviewer 2: ```suggestion
2. `loopback_ipv4_pool` defines the IP scope from which AVD assigns IPv4 addresses for Loopback0. This IP pool is identical to the one used for the spine switches in this example. To avoid setting the same IP addresses for several devices, we define the option `loopback_ipv4_offset`.
```

> 3. `loopback_ipv4_offset` offsets all assigned loopback IP addresses counting from the beginning of the IP scope. This is required when the same IP pool is used for two different node_types (like spine and l3leaf in this example) to avoid overlapping IPs. The offset is "2" in this case because each spine switch uses one loopback address.

Reviewer 2: ```suggestion
3. `loopback_ipv4_offset` offsets all assigned loopback IP addresses counting from the beginning of the IP scope. This is required when the same IP pool is used for two different node_types (like spine and l3leaf in this example) to avoid overlapping IPs. For example, the offset is "2" because each spine switch uses one loopback address.
```

4. `vtep_loopback_ipv4_pool` defines the IP scope from which AVD assigns IPv4 addresses for the VTEP (Loopback1).
5. `uplink_interfaces` defines the interfaces used locally on the leaf switches.

> 6. `uplink_switches` defines the uplink switches, which are dc1-spine1 and dc1-spine2. Note that the `uplink_interfaces` and `uplink_switches` are paired vertically.

Reviewer 3: dc2?

Author: true!

14. `node_groups` defines settings common to more than one node. For example, for leaf switches, when exactly two nodes are part of a node group, AVD will, by default, automatically generate MLAG configuration.
15. `bgp_as` is defined once since an MLAG pair shares a single BGP AS number.
16. `uplink_switch_interfaces` defines the interfaces used on the uplink switches (Ethernet1 on dc2-spine1 and dc2-spine2 in this example). Defining which spine interfaces the leaf is connected to under the leaf switch settings makes adding and removing leaf switches much easier. There is no need to add/remove any settings under the spine switch section; AVD takes care of that for you. This child device/parent device hierarchy also applies to L2 and L3 leaves.
17. `evpn_gateway` configures the EVPN DC GW features that will be inherited by the children of this group, in this case dc2-leaf2a and dc2-leaf2b. `evpn_l2` configures EVPN DC GW for EVPN type 2 routes (MAC-IP) while `evpn_l3` configures the GW for EVPN type 5 routes (IP-PREFIX).

> 18. `remote_peers` defines the RS for EVPN DC GW that will be configured on the device. This is a unique definition per device and using the hostname, AVD will be able to get all the information from the device in order to generate the configuration: Router ID to peer with and BGP AS.

Reviewer 2: ```suggestion
18. `remote_peers` defines the RS for EVPN DC GW that will be configured on the device. This is a unique definition per device, and using the hostname, AVD can get all the information from the device to generate the configuration: Router ID to peer and BGP AS.
```

> Equally, because we are adding the EVPN DC GW functionality in DC2, we need to also add it in DC1. This is a snipped part of `ansible-avd-examples/dual-dc-l3ls/group_vars/DC1.yml` file where the changes need to occour:

Reviewer 2: ```suggestion
Since we are adding the EVPN DC GW functionality in DC2, we need to add it in DC1. This is a snipped part of `ansible-avd-examples/dual-dc-l3ls/group_vars/DC1.yml` file where the changes need to occur:
```

```yaml title="DC1.yml"
    DC1_L3_LEAF2:
      bgp_as: 65102
      evpn_gateway:
        evpn_l2:
          enabled: true
        evpn_l3:
          enabled: true
          inter_domain: true
      nodes:
        dc1-leaf2a:
          id: 3
          mgmt_ip: 172.16.1.103/24
          uplink_switch_interfaces:
            - Ethernet3
            - Ethernet3
          evpn_gateway:
            remote_peers:
              - hostname: dc2-leaf2a
        dc1-leaf2b:
          id: 4
          mgmt_ip: 172.16.1.104/24
          uplink_switch_interfaces:
            - Ethernet4
            - Ethernet4
          evpn_gateway:
            remote_peers:
              - hostname: dc2-leaf2b
```

Finally, more of the same, but this time for the L2 leaf switches:

> ```yaml title="DC2.yml"

Reviewer 3: Please make title/filename 100% clear

l2leaf:
  defaults:
    platform: vEOS-lab
    uplink_interfaces: ['Ethernet1', 'Ethernet2']
    spanning_tree_mode: mstp

  node_groups:
    DC1_L2_LEAF1:
      uplink_switches: ['dc2-leaf1a', 'dc2-leaf1b']
      nodes:
        dc1-leaf1c:
          id: 11
          mgmt_ip: 172.16.1.251/24
          uplink_switch_interfaces:
            - Ethernet8
            - Ethernet8

    DC1_L2_LEAF2:
      uplink_switches: ['dc2-leaf2a', 'dc2-leaf2b']
      nodes:
        dc1-leaf2c:
          id: 12
          mgmt_ip: 172.16.1.252/24
          uplink_switch_interfaces:
            - Ethernet8
            - Ethernet8
```

As should be clear, an L2 leaf switch is much simpler than an L3 switch. Hence there are fewer settings to define.

## Specifying network services (VRFs and VLANs) in the EVPN/VXLAN fabric

> This file is identic to the one provided in the previous example and VRFs and VLANs are configured on all devices, since no `tags` or `filter` are being used.

Reviewer 2: ```suggestion
This file is identical to the one provided in the previous example. VRFs and VLANs are configured on all devices since no `tags` or `filter` are being used.
```

> It is important to consider to add the new leaves in the inventory as described in the [Inventory](#content-of-the-inventoryyml-file) section, so they will rely on this file to configure the network services.

Reviewer 2: ```suggestion
It is important to consider adding the new leaves to the Inventory as described in the [Inventory](#content-of-the-inventoryyml-file) section, so they will rely on this file to configure the network services.
```

```yaml title="NETWORK_SERVICES.yml"
--8<--

> examples/dual-dc-l3ls/group_vars/NETWORK_SERVICES.yml

Reviewer 3: Is rendering weird or is this on purpose? Seems like some content is missing?

Reviewer 2: @nielsjlarsen What does your view look like? It seems to render normally when I run mkdocs serve locally

--8<--
```

## Specifying endpoint connectivity in the EVPN/VXLAN fabric

> After the previous section, all VRFs and VLANs across the fabric are now defined accross all devices. The `ansible-avd-examples/dual-dc-l3ls/group_vars/CONNECTED_ENDPOINTS.yml` file specifies the connectivity for all endpoints in the fabric (typically servers):

Reviewer 2: ```suggestion
After the previous section, all VRFs and VLANs across the fabric are now defined on each leaf node. The `ansible-avd-examples/dual-dc-l3ls/group_vars/CONNECTED_ENDPOINTS.yml` file specifies the connectivity for all endpoints in the fabric (typically servers):
```

```yaml title="CONNECTED_ENDPOINTS.yml"
--8<--

> examples/dual-dc-l3ls/group_vars/CONNECTED_ENDPOINTS.yml

Reviewer 3: Is rendering weird or is this on purpose? Seems like some content is missing?

--8<--
```

> The configuration of the endpoints in DC2 is identic to the ones in DC1, so the only change resides in the name of the leaves used and their ports.

Reviewer 2: ```suggestion
The configuration of the endpoints in DC2 is identical to the ones in DC1. The only change resides in the name of the leaves used and their ports.
```

## The playbook

> The playbook is also the same, as the actions to execute in the fabric are the same. It is important to validate the `hosts` variable in the playbook to include all ansible groups, otherwise errors can be thrown.

Reviewer 2: ```suggestion
The playbook is also the same, as the actions to execute in the fabric are the same. It is important to validate the `hosts` variable in the playbook to include all Ansible groups. Otherwise, errors may be seen.
```
